### PR TITLE
Centralize BuildWatcher for iTunes Connect builds

### PR DESCRIPTION
**Pull request by hjanuschka**
this PR's try's to de-duplicate the various blocks that check if a build is finished processing.
by moving the logic to fastlanecore.

bildschirmfoto 2016-12-08 um 21 49 00

### CODE REVIEW

**Comment by KrauseFx** [comment]
This is great, how about we merge this and roll out a new _pilot_ release to see how it works before rolling out the change for _deliver_ and _watchbuild_? 

I didn't go through the source code yet

**Comment by hjanuschka** [comment]
@krausefx seems to be a good plan
can we have some more eyes first on the actual change in pilot+core 🤓

**Comment by hjanuschka** [comment]
updated to a single PR - should be ready for review

File: deliver/lib/deliver/submit_for_review.rb.orig
Comment (lacostej) [inline deliver/lib/deliver/submit_for_review.rb.orig]: merge issue?

**Review by lacostej — CHANGES_REQUESTED** [review]

File: deliver/lib/deliver/submit_for_review.rb.orig
Comment (lacostej) [inline deliver/lib/deliver/submit_for_review.rb.orig]: and probably lack some tests if this wasn't detected!

Comment (lacostej) [inline deliver/lib/deliver/submit_for_review.rb.orig]: I can see it, it's just a `.orig` file that shouldn't have been commited.

File: fastlane_core/lib/fastlane_core/build_watcher.rb
Comment (lacostej) [inline fastlane_core/lib/fastlane_core/build_watcher.rb]: could we get stuck here?

File: fastlane_core/lib/fastlane_core/build_watcher.rb
Comment (lacostej) [inline fastlane_core/lib/fastlane_core/build_watcher.rb]: should that method be `private`?

File: deliver/lib/deliver/submit_for_review.rb.orig
Comment (hjanuschka) [inline deliver/lib/deliver/submit_for_review.rb.orig]: ooops - thx!

File: fastlane_core/lib/fastlane_core/build_watcher.rb
Comment (hjanuschka) [inline fastlane_core/lib/fastlane_core/build_watcher.rb]: not sure if its required. - maybe there is a use case for using it public?

File: fastlane_core/lib/fastlane_core/build_watcher.rb
Comment (hjanuschka) [inline fastlane_core/lib/fastlane_core/build_watcher.rb]: seems to be, actually i moddeled it after the existing code. what is a good abort condition?
running longer than 60 minutes?

File: fastlane_core/lib/fastlane_core/build_watcher.rb
Comment (negebauer) [inline fastlane_core/lib/fastlane_core/build_watcher.rb]: Here we should put #8152

Comment (hjanuschka) [inline fastlane_core/lib/fastlane_core/build_watcher.rb]: hey @negebauer - i am unsure how to apply your pr, could you help me with this?

best would be if you checkout this branch, and modify it push it to your fork, so i can have a look.


to test run a PR/branch from a user see this here: https://gist.github.com/hjanuschka/3040fc6f371f46acff72a5e67f649b27

Comment (lacostej) [inline fastlane_core/lib/fastlane_core/build_watcher.rb]: I suggest we revisit #8152 after this one is merged, and not mix the refactoring with the bug fix.

File: fastlane_core/lib/fastlane_core/build_watcher.rb
Comment (lacostej) [inline fastlane_core/lib/fastlane_core/build_watcher.rb]: We can revisit that after this is merged.

File: fastlane_core/lib/fastlane_core/build_watcher.rb
Comment (lacostej) [inline fastlane_core/lib/fastlane_core/build_watcher.rb]: I am usually defensive when it comes to exposing code public unless needed, so I would prefer it to be private unless there's a strong argument / existing use case.

File: fastlane_core/lib/fastlane_core/build_watcher.rb
Comment (hjanuschka) [inline fastlane_core/lib/fastlane_core/build_watcher.rb]: @lacostej sounds like a solid plan ;)

File: fastlane_core/lib/fastlane_core/build_watcher.rb
Comment (hjanuschka) [inline fastlane_core/lib/fastlane_core/build_watcher.rb]: i am unsure about that.
how'd i make a instance method private?

`private_class_method :wait_for_train` - will that work on older ruby's ?

File: fastlane_core/lib/fastlane_core/build_watcher.rb
Comment (hjanuschka) [inline fastlane_core/lib/fastlane_core/build_watcher.rb]: k

File: fastlane_core/lib/fastlane_core/build_watcher.rb
Comment (negebauer) [inline fastlane_core/lib/fastlane_core/build_watcher.rb]: @lacostej Good plan

File: fastlane_core/lib/fastlane_core/build_watcher.rb
Comment (lacostej) [inline fastlane_core/lib/fastlane_core/build_watcher.rb]: 👍 

we have used `private_class_method` several times already :)

**Review by lacostej — APPROVED** [review]

**Comment by KrauseFx** [comment]
Oh yeah, this got merged, let's make sure to monitor for new issues as we roll out a new release  🚀

**Comment by hjanuschka** [comment]
@KrauseFx, @lacostej  yeah - i am bit blocked by corporate work, if you need any assitance, please direct ping me in slack, as my github notifications are a bit out of control, i think at the end of the week i am back to normal and can work a bit more on oss stuff 👍

**Pull request merged**
